routes/s3handler.js

```diff
@@ -0,0 +1,239 @@
+    };
+
+    res.setHeader("Content-Type", "application/json");
+
+    if (isPolicyValid(req.body)) {
+        res.end(JSON.stringify(jsonResponse));
+    }
+    else {
+        res.status(400);
+        res.end(JSON.stringify({invalid: true}));
+    }
+}
+
+function signV2Policy(base64Policy) {
+    return getV2SignatureKey(clientSecretKey, base64Policy);
+}
+
+function signV4Policy(policy, base64Policy) {
+    var conditions = policy.conditions,
+        credentialCondition;
+
+    for (var i = 0; i < conditions.length; i++) {
+        credentialCondition = conditions[i]["x-amz-credential"];
+        if (credentialCondition != null) {
+            break;
+        }
+    }
+
+    var matches = /.+\/(.+)\/(.+)\/s3\/aws4_request/.exec(credentialCondition);
+    return getV4SignatureKey(clientSecretKey, matches[1], matches[2], "s3", base64Policy);
+}
+
+// Ensures the REST request is targeting the correct bucket for chunking support.
+function isValidRestRequest(headerStr, version) {
+    if (version === 4) {
+        return new RegExp("host:" + expectedHostname).exec(headerStr) != null;
+    }
+
+    return new RegExp("\/" + expectedBucket + "\/.+$").exec(headerStr) != null;
+}
+
+// Ensures the policy document associated with a "simple" (non-chunked) request is
+// targeting the correct bucket and the min/max-size is as expected.
+// Comment out the expectedMaxSize and expectedMinSize variables near
+// the top of this file to disable size validation on the policy document.
+function isPolicyValid(policy) {
+    var bucket, parsedMaxSize, parsedMinSize, isValid;
+
+    policy.conditions.forEach(function(condition) {
+        if (condition.bucket) {
+            bucket = condition.bucket;
+        }
+        else if (condition instanceof Array && condition[0] === "content-length-range") {
+            parsedMinSize = condition[1];
+            parsedMaxSize = condition[2];
+        }
+    });
+
+    isValid = bucket === expectedBucket;
+
+    // If expectedMinSize and expectedMax size are not null, then
+    // ensure that the client and server have agreed upon the exact same
+    // values.
+    if (expectedMinSize != null && expectedMaxSize != null) {
+        isValid = isValid && (parsedMinSize === expectedMinSize.toString())
+            && (parsedMaxSize === expectedMaxSize.toString());
+    }
+
+    return isValid;
+}
+
+// After the file is in S3, make sure it isn't too big.
+// Omit if you don't have a max file size, or add more logic as required.
+function verifyFileInS3(req, res) {
+    function headReceived(err, data) {
+        if (err) {
+            res.status(500);
+            console.log(err);
+            res.end(JSON.stringify({error: "Problem querying S3!"}));
+        }
+        else if (expectedMaxSize != null && data.ContentLength > expectedMaxSize) {
+            res.status(400);
+            res.write(JSON.stringify({error: "Too big!"}));
+            deleteFile(req.body.bucket, req.body.key, function(err) {
```

Comment:
Can this be prevented by not uploading files that are too large in the first place?